python/cudf/cudf/core/series.py

```diff
@@ -3649,12 +3610,8 @@ def rename(self, index=None, copy=True):
         >>> renamed_series.name
         'numeric_series'
         """
-        out = self.copy(deep=False)
-        out = out.set_index(self.index)
-        if index:
-            out.name = index
-
-        return out.copy(deep=copy)
+        out = self.copy(deep=copy)
+        return Series._from_data(out._data, self.index, name=index)
```

Comment:
I'd like a specific review of this change from @vyasr or @shwina when you have time. I removed the internal use of `set_index`. Does this change use the factory methods / copy semantics properly?

Reply:
What you're doing here should have the right semantics but is making more copies than necessary. `Series._from_data` does not make copies, it just assigns attributes by reference, which is why you need to do the `self.copy` in the first line. However, we pass `self.index` rather than `out.index` to `_from_data`, which is acceptable because indices in pandas are immutable. That being the case, you could make this more efficient by only copying `self._data` (`out_data  = self._data.copy(deep=copy)`) and passing that as the first argument to `_from_data`, avoiding the unnecessary copy of the index. I would be fine making that change as part of this PR, since you're already removing (additional) unnecessary copies from the old version.

It's worth noting that `rename` is not perfectly matching pandas, which accepts [a bunch of extra arguments](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.Series.rename.html) and basically ignores them. In particular, the `copy` parameter is actually ignored by pandas:
```
>>> s = pd.Series([1, 2, 3])
>>> s1 = s.rename('a', copy=False)
>>> s1[0] = 10
>>> s
0    1
1    2
2    3
dtype: int64
>>> s1
0    10
1     2
2     3
Name: a, dtype: int64
```
Of course, we care more about memory constraints on the GPU than on the host, so maybe we're OK with this discrepancy. I'll defer to @shwina on how that should behave. Regardless of whether or not we decide to change it, any changes relating to the `rename` API are out of scope for this PR and we should make an issue documenting what we want to do going forward. Hopefully this comment is sufficient documentation for that.

Reply:
We should not create a copy unless `copy=True`.

Reply:
Thanks to both of you -- this is resolved in e4fa9ddc80e138fe984cf3a45896027c1cc78db3.